### PR TITLE
Performance improvements (+4% to speed in Magento 2.3.2)

### PR DESCRIPTION
**Pull request by andrey-legayev**
Hi,

I've run xdebug profiler for Magento2 static content deploy and found huge amount of calls to in_array() inside Less processor.
I've tried to optimize it and got ~4% performance improvement.

Pls review my changes.

Andrey

### CODE REVIEW

File: lib/Less/Tree/Selector.php [hunk 1]
```diff
@@ -108,6 +111,8 @@ public function CacheElements(){
 				array_shift($this->_oelements);
 				$this->_oelements_len--;
 			}
+
+			$this->_oelements_assoc = array_fill_keys($this->_oelements, true);
```
Comment (Krinkle) [inline lib/Less/Tree/Selector.php]: Will need to review where and how `_oelements_len` and `_oelements` are used. If they are not used anywhere else, and (especially) not cached or filled in by caches anywhere, then we might be be able to instead make this the normal form of `_oelements` itself.

File: lib/Less/Tree/Extend.php [hunk 2]
```diff
@@ -42,8 +42,10 @@ public function __construct($selector, $option, $index){
 			break;
 		}
 
-		$this->object_id = $i++;
-		$this->parent_ids = array($this->object_id);
+		// string value is needed to make sure array_merge() processes keys as strings, not as numerics
```
Comment (Krinkle) [inline lib/Less/Tree/Extend.php]: Will need to look for where this is used. Might make sense to migrate to the `+` operator which can at times be faster to do this kind of operation for associative arrays.

File: lib/Less/Visitor/processExtends.php [hunk 3]
```diff
@@ -54,7 +54,7 @@ private function doExtendChaining( $extendsList, $extendsListTarget, $iterationC
 				$targetExtend = $extendsListTarget[$targetExtendIndex];
 
 				// look for circular references
-				if( in_array($targetExtend->object_id, $extend->parent_ids,true) ){
+				if (isset($extend->parent_ids[$targetExtend->object_id])) {
```
Comment (Krinkle) [inline lib/Less/Visitor/processExtends.php]: Worth comparing to `array_key_exists`. Although counterintuitively, it seems `isset()` despite doing more work (exists + non-null), has historically been faster. Not sure if that's still the case in PHP 7.2+.

**Comment by Krinkle** [comment]
Thanks @andrey-legayev. Much appreciated.

I've filed <https://phabricator.wikimedia.org/T234270> internally to track review of this.

This Less.php fork exists mainly to keep the ship afloat to host any fixes for  incompatibilities or warnings we found during Wikipedia's migration to PHP 7.2. It doesn't currently have a long-term steward and is still marked as an "upstream" package, however that may need to change at some point - although we're short on resources in this area so that might take a while to get through. Thanks again!

File: lib/Less/Tree/Selector.php (hunk 1 above)
Comment (andrey-legayev) [reply lib/Less/Tree/Selector.php]: I tried to analyze algorithm to replace _oelements with associative array, but it's quite complicated and I've decided not to do that.

File: lib/Less/Visitor/processExtends.php (hunk 3 above)
Comment (mainframe98) [reply lib/Less/Visitor/processExtends.php]: With PHP 7.4, `array_key_exists` is now faster: https://github.com/php/php-src/pull/3360, though only when statically resolved: https://github.com/php/php-src/commit/f1c0e671e988b9bf679bd21dbf6dab9adc1e1b4e.

File: lib/Less/Tree/Selector.php (hunk 1 above)
Comment (Krinkle) [reply lib/Less/Tree/Selector.php]: Thanks. I've had a more in-depth review today, and I better understnad what you mean now.

I mistakenly thought that the array was generated elsewhere, perhaps by pushing into the array in a loop or in repsonse to other signals. But, that is not the case. It is generated right here a few lines up from `preg_match_all`. This means even if we did not keep a sequential list, we would still need to iterate the list (directly or via `array_fill_keys`). It is inevitable (without much more changes).

**Comment by andrey-legayev** [comment]
thanks for processing it!

**Pull request merged**